### PR TITLE
[WIP] Code Style

### PR DESCRIPTION
**Pull request by albertmiro**
#### 📲 What

Flutter / Dart Code Style that we will be using in the project

#### 🤔 Why
		
Simple document to highlight the practices and code style we should apply to this project.
		
#### 🛠 How
		
More in-depth discussion of the change or implementation.

#### 👀 See
		
Screenshots / external resources
		 
#### ✅ Acceptance criteria

- [ ] Design Review for UI with BA completed. 
- [ ] Manually tested and verified on Android device/emulator (min. Lollipop 5.1)
- [ ] Showcase video / automation test
- [ ] Passing all tests
- [ ] Rebased/merged with latest changes from development and re-tested
- [ ] Removed unsed comments. TODOs must have JIRA for future work.

#### Coding Standards Checklist
- [ ] unit tests 80% coverage on testable code (functions, methods, class)
- [ ] (Architecture) redux state, reducers, actions, middleware defined under lib/redux
- [ ] ui elements defined under lib/ui. 
- [ ] Strings ready for localisation (e.g. defined in lib/utils/strings.dart)
- [ ] Assets (images/icons path) defined in lib/utils/assets.dart
- [ ] Colors defined in lib/utils/colors.dart
- [ ] Dimensions (ie margins padding) defined in lib/utils/dimens.dart
- [ ] TextStyles defined in lib/utils/styles.dart

Recommended Style Guide: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo

#### 🕵️‍♂️ How to test

Notes for QA

#### Reviewers

@farmsmart/amido @farmsmart/wamf

### CODE REVIEW

**Review by maribel4amido — DISMISSED** [review]

**Comment by LeeMatthewHiggins** [comment]
lets get the PR template into this code style PR too.

File: docs/code-style.md [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+## Strings
+
+The same that we said with Constants applies with Strings.
+
+In order to keep our resources clean and organized, we would separate the Strings by their use.
+
+Let's see an example:
+
+```
+class Strings {
```
Comment (LeeMatthewHiggins) [inline docs/code-style.md:83]: Here we don't want global scope. so things called strings are only valid if file private. Constant should be on a widget basis, else when you try and reuse code you will have to lift out all the constants.

File: docs/code-style.md [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    String myProfileDescription = ""
+    String myProfileDate = ""
+    String myProfileSkills = ""
+}
+```
+We should refactor that an have:
+
+```
+class Strings {
+    class BottomBarTabs {
```
Comment (LeeMatthewHiggins) [inline docs/code-style.md]: can dart have inner classes? not sure it can (at least at one point you couldn't)

File: docs/code-style.md [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        String home = ""
+        String profile = ""
+        String news = ""
+        String more = ""
+    }
+
+    class MyProfile{
```
Comment (LeeMatthewHiggins) [inline docs/code-style.md]: almost, we have context on the constants 👍 , but we are mixing bottom bar and profile. in this file. If I only want bottom bar I will have to extract profile.

File: docs/code-style.md [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+The architecture approach we took in this project is to use **Redux** as the main architecture to follow to separate all the app layers.
+
+## Enum
+
+When we have a variable that can take different states, we will use an *Enum* over integers
+
+Here's an example, of using constants vs Enum values for defining complexity:
+
+```
+const easy = 0;
```
Comment (LeeMatthewHiggins) [inline docs/code-style.md:16]: we should have a rule, no global scope constants 👍

**Review by LeeMatthewHiggins — CHANGES_REQUESTED** [review]

File: docs/code-style.md (hunk 2 above)
Comment (albertmiro) [reply docs/code-style.md]: Ups, yes, I didn't know that, I will change this part to follow the rules we defined and don't have the global scope Strings file.

File: docs/code-style.md (hunk 3 above)
Comment (albertmiro) [reply docs/code-style.md]: Yep, that was with the old approach, I will change it to reflect that should be private for the Widget

File: docs/code-style.md [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+- [ ] Showcase video / automation test
+- [ ] Passing all tests
+- [ ] Rebased/merged with latest changes from development and re-tested
+- [ ] Removed unsed comments. TODOs must have JIRA for future work.
+
+#### Coding Standards Checklist
+- [ ] unit tests 80% coverage on testable code (functions, methods, class)
+- [ ] (Architecture) redux state, reducers, actions, middleware defined under lib/redux
+- [ ] ui elements defined under lib/ui. 
+- [ ] Strings ready for localisation (e.g. defined in lib/utils/strings.dart)
```
Comment (LeeMatthewHiggins) [inline docs/code-style.md]: remove "(e.g. defined in lib/utils/strings.dart)" as this is implementation specific

File: docs/code-style.md [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+- [ ] Assets (images/icons path) defined in lib/utils/assets.dart
+- [ ] Colors defined in lib/utils/colors.dart
```
Comment (LeeMatthewHiggins) [inline docs/code-style.md]: remove this, we have style injection and then app specific styles will be in an app scope not a global scoped generic file.

File: docs/code-style.md [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+- [ ] Dimensions (ie margins padding) defined in lib/utils/dimens.dart
```
Comment (LeeMatthewHiggins) [inline docs/code-style.md]: same as above, the dim´s are in the style injection

File: docs/code-style.md [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+- [ ] TextStyles defined in lib/utils/styles.dart
```
Comment (LeeMatthewHiggins) [inline docs/code-style.md]: same as above

File: docs/code-style.md [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+		 
+#### ✅ Acceptance criteria
+
+- [ ] Design Review for UI with BA completed. 
+- [ ] Manually tested and verified on Android device/emulator (min. Lollipop 5.1)
+- [ ] Showcase video / automation test
+- [ ] Passing all tests
+- [ ] Rebased/merged with latest changes from development and re-tested
+- [ ] Removed unsed comments. TODOs must have JIRA for future work.
+
```
Comment (LeeMatthewHiggins) [inline docs/code-style.md:218]: lets add
[ ] Snap shot test for new widgets
[ ] Unit test if in Jira ticket

**Review by LeeMatthewHiggins — CHANGES_REQUESTED** [review]

File: docs/code-style.md (hunk 5 above)
Comment (albertmiro) [reply docs/code-style.md]: Changed these points.
Updated the architecture, due that we are moving to BLOC.
Added:
- [ ] No global scope strings, dimens, styles or colors
- [ ] Using style injection for Widgets

Comment (LeeMatthewHiggins) [reply docs/code-style.md]: 👍

**Review by LeeMatthewHiggins — APPROVED** [review]

**Review by isaacmartinezimv — APPROVED** [review]

**Review by davidfranquet — APPROVED** [review]

**Pull request merged**